pkg/metadata/host/host.go

```diff
@@ -152,6 +152,26 @@ func getHostAliases() []string {
 	return aliases
 }
 
+// try to get the public IPv4 address using AWS and GCE
+func getPublicIPv4() (string, error) {
+	publicIPv4, err := ec2.GetPublicIPv4()
+	if err == nil {
+		log.Debugf("EC2 public IP = %s", publicIPv4)
+		return publicIPv4, err
+	}
+	log.Debugf("no EC2 public IPv4: %s", err)
+
+	publicIPv4, err = gce.GetPublicIPv4()
+	if err == nil {
+		log.Debugf("GCE public IP = %s", publicIPv4)
+		return publicIPv4, err
+	}
+	log.Debugf("no GCE public IPv4: %s", err)
+
+	log.Infof("No public IPv4 address found")
+	return "", err
+}
```

Comment:
Could we make it simpler to add other cloud provider in the futur.

Something like:
```suggestion
func getPublicIPv4() (string, error) {
	publicIPFetcher := map[string]func()(string,error) {
		"ec2": ec2.GetPublicIPv4(),
		"gce": gce.GetPublicIPv4(),
	}

	for name, fetcher := range publicIPFetcher {
		if publicIP, err := fetcher(), err == nil {
			log.Debugf("%s public IP = %s", name, publicIP)
			return publicIPv4, nil
		}
		log.Debugf("could not fetch %s public IPv4: %s", name, err)
	}

	log.Infof("No public IPv4 address found")
	return "", nil
}
```

Reply:
Seems like a more complex way to do the same thing - what I really want is a "which cloud are we in" that I can use to just call the single function appropriately.

Reply:
My example allows to modify only 1 line when we will add other cloud provider and remove copy/pasting the same code snippet for each cloud.

For the overall detection of the cloud provider, it's something we have in our backlog but will require some deeper rewrite.

Reply:
Since you are an owner of the codebase, I will happily defer to you on how you want the code written. Since I find such things interesting, let me lay out my thoughts on the matter:

I find myself these days looking at the code through the eyes of the "Future Person", and since it is usually [harder to read code than write it](https://www.joelonsoftware.com/2000/04/06/things-you-should-never-do-part-i/), my general goal is to help the Future Person understand the code they are modifying.

If we look at the two versions of the code, with the "repeated if" version, the actual code that needs to be written is a function call, an if checking the err code, and a return. With the "function map" version, it is only one line of code: adding adding to a map. Just looking at that, I agree the "function map" version is simpler to modify (slightly!)

However, if you look at it with the lens that a person coming to this code hasn't seen it before, reading the two versions of the code is a different experience. With the "function map" version, Future Person comes in and sees a map, which is common, but this one has a "function" in it, which is much less common. So now they are wondering maybe something complex is going on here, and so they read the rest of the for loop, and work to make sure they aren't missing something. Only then do they add the line that is needed. Unless, of course, what they are adding needs a slightly different way of handling, then they need to refactor the whole thing.

With the "repeated if" version, it's a series of if-then statements, using very basic features of Go. It is, in fact, rather boring code. Boring usually means easy to understand, which is exactly what we want to help Future Person make the correct decision simply and quickly and confidently. 

Or at least that's my hypothesis :)

Anyway, as I said, if you don't find that spiel about a tiny bit of code from a grumpy old engineer persuasive, I'm 100% happy to redo it!

Reply:
I understand you point of view but I still disagrees.

First a map of functions is very basic and should not throw off the "Future Person".

Then, and it's mostly the main advantage I see, duplication is error prone for "Future Person". The more code duplication we have the easier it is to only change a part, forget a case and so on. It's also easier to get the big picture of what the function is trying to do. It's upon modification that we introduce bugs, so the simpler the edit the better.

Lastly, we want to have all cloud providers behave the same. It's there job to encapsulate any differences. That's why I think it's a good thing we have a generic usage of them. Having a if/else for each opens the door for exception while we want to have a generic code.